File: LL/src/Presentation/ll/src/app/shared/models/soulstones/soulstone-upgrade-view.ts
export type SoulstoneUpgradeBranch =
  | 'EssenceArchive'
  | 'CombatProgression'
  | 'Gathering'
  | 'Crafting'
  | 'Dungeons'
  | 'AccountConvenience';

export interface SoulstoneUpgradeView {
  id: string;
  branch: SoulstoneUpgradeBranch;
  displayName: string;
  description: string;
  currentRank: number;
  maxRank: number;
  currentEffectText: string;
  nextEffectText?: string | null;
  nextCost?: number | null;
  canPurchase: boolean;
  disabledReason?: string | null;
  appliesTo: string[];
  doesNotApplyTo: string[];
  refundValue: number;
  sortOrder: number;
  frontendHint?: string | null;
}

export interface SoulstoneUpgradeMutationResult {
  upgrades: SoulstoneUpgradeView[];
  soulstones: number;
  refundedSoulstones: number;
}

File: LL/src/Core/Domain/Models/Soulstones/UpgradeDefinition/SoulstoneUpgradeView.cs
namespace Domain.Models.Soulstones.UpgradeDefinition;

public sealed record SoulstoneUpgradeView(
    string Id,
    SoulstoneUpgradeBranch Branch,
    string DisplayName,
    string Description,
    int CurrentRank,
    int MaxRank,
    string CurrentEffectText,
    string? NextEffectText,
    int? NextCost,
    bool CanPurchase,
    string? DisabledReason,
    IReadOnlyList<string> AppliesTo,
    IReadOnlyList<string> DoesNotApplyTo,
    int RefundValue,
    int SortOrder,
    string? FrontendHint);
